LE Audio: Fixes issues and race conditions when disconnecting CIS

Thalley opened the pull request:
Fixes multiple issues with disconnecting the CIS in the unicast client. 

This PR adds additional state checks and proper handling of disconnection of CIS when two endpoints are involved. 

Tested with https://github.com/zephyrproject-rtos/zephyr/pull/54441 which discovered the issue

fixes https://github.com/zephyrproject-rtos/zephyr/issues/54538
fixes https://github.com/zephyrproject-rtos/zephyr/issues/54539

Thalley committed:
Bluetooth: Audio: Unicast client check ISO state before disconnect

Add a check to see if the ISO is connected before we attempt to
disconnect it. In the case where a stream was not in the streaming
state and was released, the CIS is unlikely to be connected.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Fix state checks for bt_iso_chan_disconnect

The behavior of bt_iso_chan_disconnect depends on the
state of the ISO channel.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add addition ISO chan state check before disconnect

Add state check for BT_ISO_STATE_CONNECTING before disconnecting
the CIS, else we could end up in a race condition where we
did not disconnect the connecting CIS, and then still have
a CIS afterwards.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Handle GATT/ISO race condition in unicast client

When releasing a stream causing the CIS to disconnect, there was
a race condition between the CIS disconnect event and the
idle state notification from the unicast server.

This ensures that if we are disconnecting a CIS as part of the
release, then we wait for the CIS disconnect event before calling
the "released" callback.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

MariuszSkamra commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -730,12 +730,15 @@ static void unicast_client_ep_releasing_state(struct bt_audio_ep *ep,
-	 * Client has determined that the ASE is in the Releasing state.
-	 */
-	bt_audio_stream_disconnect(stream);
+	if (ep->iso != NULL && ep->iso->chan.state == BT_ISO_STATE_CONNECTED) {
```
Shouldn't disconnect be called as well if `BT_ISO_STATE_CONNECTING` like you did in other places? Going from Enabling to Releasing I guess there might be no CIS Connected yet (looking at Figure 3.2 in ASCS)

Thalley replied:
The addition check for `BT_ISO_STATE_CONNECTING` is added an a later commit in this PR :) At the time of this commit, no other places checks it.

Thalley committed:
Bluetooth: Audio: Ensure stream disconnect correctness for unicast client

Add a new function for the unicast client that verifies whether
or not the unicast client can terminate the CIS given the
state of any paired endpoint and the CIS state.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

laurenmurphyx64 commented:
Hi @Thalley, link a bug report for this to be merged for 3.3
